fmpz/mul_2exp.c

```diff
@@ -9,38 +10,103 @@
     (at your option) any later version.  See <https://www.gnu.org/licenses/>.
 */
 
-#include <gmp.h>
-#include "flint.h"
-#include "ulong_extras.h"
 #include "fmpz.h"
+#include <string.h>
 
 void
 fmpz_mul_2exp(fmpz_t f, const fmpz_t g, ulong exp)
 {
-    fmpz d = *g;
+    slong c1 = *g;
+    ulong c1abs, c1bits;
+    __mpz_struct * mf;
 
-    if (!COEFF_IS_MPZ(d))       /* g is small */
+    if (c1 == 0)
     {
-        ulong dabs = FLINT_ABS(d);
-        ulong bits = FLINT_BIT_COUNT(dabs);
-        if (bits == 0)
+        fmpz_zero(f);
+        return;
+    }
+
+    c1abs = FLINT_ABS(c1);
+    c1bits = FLINT_BIT_COUNT(c1abs);
+
+    if (c1bits + exp <= FLINT_BITS - 2)  /* Result fits inside a small fmpz */
+    {
+        if (COEFF_IS_MPZ(*f))
+            _fmpz_clear_mpz(*f);
+
+        *f = c1 << exp;
+    }
+    else if (c1bits <= FLINT_BITS - 2)   /* g is small */
+    {
+        ulong expred = exp % GMP_NUMB_BITS;
+        int alloc = 1 + (FLINT_BITS != GMP_NUMB_BITS) + exp / GMP_NUMB_BITS
+            + ((c1bits + expred) > GMP_NUMB_BITS);
+        mp_limb_t * limbs;
+        slong tmp;
+
+        /* Ensure enough limbs are allocated for f */
+        if (!COEFF_IS_MPZ(*f))
         {
-            fmpz_set_si(f, 0);
+            /* TODO: Initialize the new mpz with alloc limbs instead of
+             * reallocating them. */
+            mf = _fmpz_new_mpz();
+            *f = PTR_TO_COEFF(mf);
+            _mpz_realloc(mf, alloc);
         }
-        else if (bits + exp <= FLINT_BITS - 2)  /* result will fit in small */
+        else
         {
-            fmpz_set_si(f, d << exp);
+            mf = COEFF_TO_PTR(*f);
+            if (mf->_mp_alloc < alloc)
+                _mpz_realloc(mf, alloc);
         }
-        else                    /* result is large */
+        limbs = mf->_mp_d;
+        mf->_mp_size = (c1 > 0) ? alloc : -alloc;
+        memset(limbs, 0, sizeof(mp_limb_t) * (alloc - 1));
```

Comment:
If you don't just want to use mpz_mul_2exp here, then flint_mpn_zero is better (more idiomatic) than memset.

Reply:
Seconded. Ignore my comment here as this is a better one.

Reply:
Even though one might not often use large numbers, I do think that memset always will be faster than flint_mpn_zero. Or is the compiler smart enough to always convert flint_mpn_zero into memset?

Reply:
mpn_zero should be faster. I don't recall why we have flint_mpn_zero. Likely because the former doesn't deal with zero length correctly on Windows or something. Certainly GMP has a very fast implementation of this function.

Reply:
mpn_zero makes sense if you need to zero huge arrays. For a function like this likely to be used on 1-2 limbs on average it's probably not worth the function call overhead.

Anyway, compilers can recognize for loops that do the equivalent of a memset and may generate vectorized code on architectures where that makes sense.

Reply:
I will stick with memset unless someone gets problems with using it as it can be inlined on many systems.  (It is already used under `fmpz_[factor/mpoly]` and `qsieve`, but I'll change it if you really don't want it.)

Reply:
It's used in qsieve because the sieve being allocated is an array of bytes and is quite large.

If you use memset, please deal with the case of zero limbs as this will break on some systems. I agree with Fredrik this will be much slower for small limb counts and should not be used in this case. You can use it if you want, but I don't think this is the right situation to do so.

Fredrik is right that modern compilers automatically vectorise when the limb count is high (and constant). So flint_mpn_zero may actually be faster than memset in many cases.

Reply:
I already dealt with the zero case! :-)